### PR TITLE
Extend `Print Canonical Projections` with a search functionality

### PR DESCRIPTION
**Pull request by pi8027**
To see how canonical projections work and which projection/pattern they involve in advanced canonical structure programming, we need to use the `Print Canonical Projections` command and find a unification rule from a bunch of output lines. This PR allows `Print Canonical Projections` to take constants and print only the unification rules that involve or are synthesized from given constants.

**Kind:** feature.

- [x] Added / updated test-suite

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: pretyping/recordops.ml [hunk 1]
```diff
@@ -246,13 +246,23 @@ let warn_redundant_canonical_projection =
...
 
+let canonical_projections_printer :
+  (Environ.env -> Evd.evar_map -> ((GlobRef.t * cs_pattern) * obj_typ) list -> Pp.t) ref =
+  ref (fun _ _ _ -> str "")
+
+let install_canonical_projections_printer f = canonical_projections_printer := f
+
+let print_canonical_projections env sigma cs = !canonical_projections_printer env sigma cs
+
 let register_canonical_structure ~warn env sigma o =
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply pretyping/recordops.ml]: Indeed that's the wrong place to add this print statement.

**Review by ejgallego — DISMISSED** [review]
I don't think the core of Coq should be noisy by default, in particular printing what amounts to debug information. If a user needs to see info they can always call the corresponding `Print` command, hover on the UI, etc...

File: pretyping/recordops.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (pi8027) [reply pretyping/recordops.ml]: Yes. I didn't expect this behavior when I start this.

**Comment by pi8027** [comment]
@ejgallego So could we have a new option to control this verbosity, or change the `About` command to print canonical projections associated with the given constant? (But `object_table` in `recordops.ml` isn't designed in this way so now I don't see how to do it.)

**Comment by SkySkimmer** [comment]
We can always add a table if it's useful

Gaëtan Gilbert

**Comment by ejgallego** [comment]
Yeah improving the table/`About` command seems right.

**Comment by pi8027** [comment]
@SkySkimmer @ejgallego I will do so. Thanks!

**Comment by pi8027** [comment]
I have almost finished the implementation, but I'm still not sure about my design. I will discuss with someone to make sure of it. (Maybe I need to change `Print Canonical Projection` to take a constant optionally rather than changing the `About` command.)

**Comment by pi8027** [comment]
Since I have no idea on how this should work with PR #10984 (declaring multiple instances at once), I would like to have comments from @gares and @CohenCyril.

**Comment by gares** [comment]
If you want to scale this to the other pr you should make the other pr keep a link between what the user wrote and the derived cs instances I'm afraid.

**Comment by pi8027** [comment]
So IIUC, I should extend the object table ‘obj_typ’ with a gref. Using ‘o_DEF’ for this purpose seems to be a bad idea anyway.

**Comment by gares** [comment]
Maybe o_ORIGIN : gref option. But you should not wait for the other pr, this change morally belongs to the other one.

**Comment by pi8027** [comment]
@gares Since I didn't see why you prefer `option`, I put `o_ORIGIN : Constant.t`.

**Comment by gares** [comment]
It is OK.

I was proposing option because for regular CS you have o_DEF = o_ORIGIN, and you have != for multiple CS. With option you could have None for the first case, and (Some origin) for the second.
Your code tests for equality, while it could match on the option... not very different.

**Comment by herbelin** [comment]
Looks relevant and very good to me. (But I must confess that I did not understand @ejgallego's objection about noisiness. Isn't the PR only allowing to reduce the verbosity of the `Print Canonical Projections` query?)

**Comment by ejgallego** [comment]
> But I must confess that I did not understand @ejgallego's objection about noisiness. Isn't the PR only allowing to reduce the verbosity of the `Print Canonical Projections` query?

My comment was referring to the original proposal of print the projections after each `Canonical` commands, I think the current proposal doesn't do that so my review can be disregarded.

**Comment by pi8027** [comment]
@ejgallego @herbelin Yes. I changed the design of this feature twice. Old implementations were discarded but they can be seen in the changes of the title.

I have just added a comment and rebased again. I think this is ready to merge if CI succeeds.

File: doc/sphinx/language/gallina-extensions.rst [hunk 2]
```diff
@@ -2075,11 +2075,13 @@ in :ref:`canonicalstructures`; here only a simple example is given.
...
 
-.. cmd:: Print Canonical Projections
+.. cmd:: Print Canonical Projections {* @ident}
 
    This displays the list of global names that are components of some
    canonical structure. For each of them, the canonical structure of
-   which it is a projection is indicated.
+   which it is a projection is indicated. If constants are given as
+   its arguments, only the unification rules that involve or are
+   synthesized from given constants will be shown.
```
Comment (CohenCyril) [inline doc/sphinx/language/gallina-extensions.rst]: I suggest the following
```suggestion
   synthesized from simultaneously all given constants will be shown.
```
to disambiguate that all constraints must be satisfied and not at least one.

**Review by CohenCyril — CHANGES_REQUESTED** [review]
The feature, documentation and test suite together with its output all look good and useful to me.
The ocaml code is clean enough for me to read and understand.

I am all in favor of merging up to the small remark I am making, and then as soon as it passes CI.

**Comment by gares** [comment]
@pi8027 if you rebase and implement the fix requested by Cyril I'll happily merge (CI failure is unrelated)

**Comment by pi8027** [comment]
@gares Done.

**Review by CohenCyril — APPROVED** [review]

**Review by ejgallego — APPROVED** [review]

**Pull request merged**